ExampleCode/C++/Apps/API.cpp

```diff
@@ -0,0 +1,1618 @@
+    if(http_code == 422)
+    {
+      cerr << "Something went wrong with iSENSE.\n";
+      cerr << "Try formatting your data differently,\n";
+      cerr << "using a contributor key instead of an email/password,\n";
+      cerr << "or asking for help from others. You can also try running the\n";
+      cerr << "the program with the \"debug\" method enabled, by typing: \n";
+      cerr << "object_name.debug()\n";
+      cerr << "This will output a ton of data to the console and may help you in\n";
+      cerr << "debugging your program.\n";
+    }
+    if(http_code == CURL_ERROR)
+    {
+      cerr << "\n\nCurl failed for some unknown reason.\n";
+      cerr << "Make sure you've installed curl / libcurl, and have the \n";
+      cerr << "picojson header file as well.\n";
+    }
+  }
+
+  return false;
+}
+
+
+/*
+ *    Appends to a dataset using its dataset name, which can be used to find a dataset ID
+ *    We can find the dataset ID by comparing against all the datasets in a given project
+ *    until we find the dataset with the given name.
+ *
+ */
+bool iSENSE::append_email_byName(string dataset_name)
+{
+  // Check that the project ID is set properly.
+  // When the ID is set, the fields are also pulled down as well.
+  if(project_ID == "empty" || project_ID.empty())
+  {
+    cerr << "\nError - please set a project ID!\n";
+    return false;
+  }
+
+  // Check that a title and contributor key has been set.
+  if(title == "title" || title.empty())
+  {
+    cerr << "\nError - please set a project title!\n";
+    return false;
+  }
+
+  if(email == "email" || email.empty())
+  {
+    cerr << "\nErrror - please set an email address!\n";
+    return false;
+  }
+
+  if(password == "password" || password.empty())
+  {
+    cerr << "\nErrror - please set a password!\n";
+    return false;
+  }
+
+  // Make sure the map actually has stuff pushed to it.
+  if(map_data.empty())
+  {
+    cerr << "\nMap of keys/data is empty. You should push some data back to this object.\n";
+    return false;
+  }
+
+  // We can now find the dataset ID by comparing against all the datasets in this project.
+  // First pull down the datasets
+  get_datasets_and_mediaobjects();
+
+  // Call the get_dataset_ID function
+  string dataset_ID = get_Dataset_ID(dataset_name);
+
+  // If we didn't get any errors, call the append by ID function.
+  if(dataset_ID != GET_ERROR)
+  {
+    append_key_byID(dataset_ID);
+    return true;
+  }
+
+  // If we got here, we failed to find that dataset name in the current project.
+  cerr << "Failed to find the dataset name in project # " << project_ID << endl;
+  cerr << "Make sure to type the exact name, as it appears on iSENSE. \n";
+  return false;
+}
+
+
+// This function is called by the JSON upload function
+// It formats the upload string
+// Users should not have to call this function - API methods will,
+// and will pass an int value indicating which API method they are using.
+void iSENSE::format_upload_string(int key)

```

Comment:
That's a good idea. I use an int to determine which post method will be called. This allows me to use one method four times (Contributor Key, Email, and appending by both). The "key" was originally a bool I think, and was for whether it was a post using keys or email/password. A better name might be "post_type" or such.